banking/src/main/java/rocks/kata/Account.java

```diff
@@ -0,0 +1,61 @@
+package rocks.kata;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.LongAdder;
+
+import rocks.kata.AccountOperation.OperationType;
+import rocks.kata.DatabaseClasses.AccountOperationsDatabase;
+
+public class Account {
+
+    private static final LongAdder numberOfAccounts = new LongAdder();
+
+    private AtomicInteger balance;
+    private long uniqueAccountID;
+    private AccountOperationsDatabase operationsDatabase;
+
+    public Account(AccountOperationsDatabase operationsDatabase) {
+        numberOfAccounts.increment();
+        this.uniqueAccountID = numberOfAccounts.longValue();
+        this.balance = new AtomicInteger(0);
+        this.operationsDatabase = operationsDatabase;
+        saveOperationToDatabase(OperationType.OPENING, 0);
+
+    }
+
+    public void withdraw(int amount) throws NotEnoughBalanceException {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("The transaction value must be positive number");
+        }
+        if (balance.intValue() - amount < 0) {
```

Comment:
To nie chroni przed błędami na wielu wątkach.
1. Wątek A pobiera wartosc = 10.
2. Watek A wylicza roznice = 3 (amount = 7).
3. Watek B pobiera wartosc = 10.
4. Watek A sprawdza, ze 3 nie jest mniejsze od 0.
5. Watek B wylicza roznice = 4 (amount = 6).
6. Watek B sprawdza, ze 4 nie jest mniejsze od 0.
7. Oba watki wykonuja operacje => koncowy stan konta to -3.

Reply:
Ok, widzę to tak, że zamykam kod od linijki 30 do 33 w bloku synchronized, bo:
 - każda z atomowych metod AtomicIntegera zmienia mi wartość, a nie zawsze chcę zmieniać wartość `balance`
 - metoda `saveOperationToDatabase' woła metodę calculateBalanceAfterOperation` i  znowu pobiera wartość `balance`, więc trzeba zapewnić że to ten sam `balance` (zauważyłam też że w sumie w `saveOperationToDatabse` robiłam operacje na `balance` trzy razy, zredukowałam to teraz do jednej operacji):

![image](https://user-images.githubusercontent.com/93493337/217944109-ceb12a2b-35fa-4d4b-9df8-7450dddc5013.png)

Metoda deposit będzie wołała tylko raz operację na `balance` więc jej nie ruszam.

Co o tym myślisz?

Reply:
Yup, dokładnie o `synchronized` wokół kodu, który operuje na jednym polu, do którego dostęp jest wielowątkowy, chodzi :)